### PR TITLE
[Kernel] Triton-based Top-k and Top-p sampler kernels

### PR DESCRIPTION
**Pull request by cakeng**
## Purpose
The current method for top-k and top-p sampling utilizes a sorting-based approach on PyTorch operations, which is suboptimal.

This PR introduces pivot-search-based top-k and top-p sampling kernels using Triton, which achieves a speedup of up to 10 times compared to the native method.

Set the env variable "VLLM_USE_TRITON_SAMPLER=1" to enable the Triton-based top-k and top-p sampler.

## Test Plan
Accuracy and speedup of the Triton kernels are tested against the native kernel using this [script](https://github.com/cakeng/vllm/blob/9b3cf75aacde92b8978aad3752c6a54ff0b60e4a/test_triton_topk_topp.py).

### Speed
Full test results are available [here](https://docs.google.com/spreadsheets/d/15CSXj_Bi9RMBEekPy8qBM7O3oBjdwTOa-NTCKRPIPWo/edit?usp=sharing).
Triton_topk_topp_speedup_analysis_H100
(RAND results randomly select p values between 0.0 and 1.0, and k values between 1 and vocab_size, for each batch index.)

The Triton-based kernel works better with larger batch sizes and K < vocab_size * 0.03. As modern LLMs have a vocab_size of around 128k, the kernel works best when K < 4000.

Each Triton program in the kernel processes a single batch index of the logits tensor, meaning that the kernels will not fully utilize the given hardware when the batch size is less than the number of SMs in the GPU. As such, the kernel works better with larger batch sizes.

The kernel also contains a "outlier search" optimization where it will gather the logits which are larger than the avg(logit) + 2.15 *std_dev(logit), which is about 3% of the vocab size, assuming a Gaussian distribution. If the number of outliers is larger than K, or if the sum of the outlier probabilities is larger than P, the kernel will only search within the outlier set, which greatly reduces search time. The coefficient for the standard deviation, 2.15, was randomly chosen. A better algorithm where the coefficient is dynamically adjusted based on the ratio of k and vocab_size may be introduced to further improve the execution.

**Speed TLDR** 
While the kernels are overall faster than the naive Pytorch implementations, they are slower on smaller batch sizes, or when K or P is large enough that the number of returned logits are over vocab_size*0.03.

### Accuracy

There exist cases where the output of the Triton implementation does not match that of the Pytorch kernel, especially for Top-p. 

The pivot-based search algorithm will not work correctly if the pivot falls on a duplicate value. For example, pivot-based top-3 on an array of [3, 3, 2, 2, 1] will fail. The top-k kernel implementation is designed to return all duplicate values if this happens, mimicking that of the existing "apply_top_k_only()" method. 

For top-p, the problem is slightly more complex, as there are more duplicate entries, and the native Pytorch implementation does not simply return all duplicate entries. By definition, top-p must return the smallest set of logits whose probabilities add up to or exceed p. The Softmax function amplifies the difference between logits with a large difference. However, this also means that the differences between similar logits are reduced after Softmax, and many of the smaller logits are mapped to the same probability.

The native Pytorch approach applies simple sorting on the logits and cumulative sum on the probability to cut off at a certain probability value. This means that even if there are multiple duplicate probabilities, the first n probabilities that tip the cumulative sum over p will be selected.  As such, the native Pytorch implementation will select the ones that were mapped to the earlier index in the sorted list by torch.sort among the duplicate probabilities.

To mimic this behavior, the Triton top-p kernel will keep track of duplicate entries near the pivot. If it determines that some portion of the duplicate entries must be included to hit top-p, it will manually include those in the final output. HOWEVER, the inclusion is decided by the index order of the original logit position, AND NOT the sorted logit position. In this case, the Triton-based top-p kernel WILL return a different result from the native implementation.

**Accuracy TLDR** 
Top-k will return the same result as the native top-k implementation, which may return more than k values. Top-p will return the results that satisfy the top-p condition, but may return a different set of values compared to the Pytorch implementation. Both kernels will fail when all logits are same, or if they contain inf, -inf, or NaN.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/v1/sample/ops/topk_topp_sampler.py [hunk 1]
```diff
@@ -204,6 +234,636 @@ def apply_top_k_top_p(
...
+                                       vocab_size)
+    elif k is None and p is not None:
+        probs = torch.full_like(logits, -float('inf'), device=logits.device)
+        probs_2 = torch.full_like(logits, -float('inf'), device=logits.device)
+        _topp_kernel[(NUM_PROGRAMS, )](logits, probs, probs_2, p, batch_size,
+                                       SIGMA, vocab_size)
+    elif k is not None and p is not None:
+        probs = torch.full_like(logits, -float('inf'), device=logits.device)
+        _topk_topp_kernel[(NUM_PROGRAMS, )](logits, probs, k, p, batch_size,
+                                            SIGMA, vocab_size)
```
[the remark this answers is not in the document]
Comment (cakeng) [reply vllm/v1/sample/ops/topk_topp_sampler.py]: Smart bot!

**Review by WoosukKwon — COMMENTED** [review]
@cakeng Huge thanks for the PR, and sorry for the late review.

I'm looking into more details, but high-level comments:
1. I like the idea! Very interesting. 🔥 
2. Can you please add a test to check the correctness?
3. Can we refactor the top_p and top_k kernels so that we don't need to duplicate the code in `_topk_topp_kernel`? I think we can just simply define the core logic as a separate function and call it in two kernels.
4. Can you please benchmark the performance against the FlashInfer kernel as well? The kernel doesn't have to be faster, but it'd be helpful to know the gap. I think a realistic setup is `top_p=0.95` and `top_k=20` as in https://huggingface.co/Qwen/Qwen3-VL-30B-A3B-Thinking/blob/main/generation_config.json

File: vllm/v1/sample/ops/topk_topp_sampler.py [hunk 2]
```diff
@@ -213,6 +242,646 @@ def apply_top_k_top_p(
...
+                                            SIGMA,
+                                            vocab_size,
+                                            BLOCK_SIZE,
+                                            num_warps=NUM_WARPS,
+                                            num_stages=NUM_STAGES)
+    return logits
+
+
+@triton.jit
+def _topk_kernel(LOGITS, PROBS, K, B, SIGMA: tl.constexpr, N: tl.constexpr,
```
Comment (WoosukKwon) [inline vllm/v1/sample/ops/topk_topp_sampler.py]: nit:
```suggestion
def _topk_kernel(LOGITS, PROBS, K, B, SIGMA: tl.constexpr, VOCAB_SIZE: tl.constexpr,
```

File: vllm/v1/sample/ops/topk_topp_sampler.py [hunk 3]
```diff
@@ -213,6 +242,646 @@ def apply_top_k_top_p(
...
+                 BLOCK_SIZE: tl.constexpr):
+    pid = tl.program_id(0)
+    num_programs = tl.num_programs(0)
+    NUM_TILES: tl.constexpr = (N + BLOCK_SIZE - 1) // BLOCK_SIZE
+    for row_id in tl.range(pid, B, num_programs):
```
Comment (WoosukKwon) [inline vllm/v1/sample/ops/topk_topp_sampler.py]: QQ: Why do we use the persistent kernel technique here? Does this improve performance? Other wise, I feel like we can simplify the kernels quite a bit by not using the persistent style.
